File: c++/321. Create Maximum Number.cpp
class Solution {
public:
    // 贪心求数组中的最大的大小为k的子数组 
    void maxkseq(vector<int> &a, vector<int> &res, int k) {
        if (k == 0) return;
        int n = a.size();
        int i, j;
        res.resize(k, 0);
        
        for (i = 0, j = 0; i < n; i++) {
            while (j > 0 && n-i+j > k && a[i] > res[j-1]) // 出栈
                j--;
            if (j < k) // 入栈
                res[j++] = a[i];        
        }
    }
    
    void merge(const vector<int>& v1, const vector<int>& v2, vector<int> &v3) {
        int lv1 = v1.size(), lv2 = v2.size(), lv3 = lv1 + lv2;
        v3.resize(lv3, 0);
        int i, j, k;
        i = j = k = 0;
        while (i < lv1 || j < lv2) {
            v3[k++] = compare(v1, v2, i, j) ? v1[i++] : v2[j++];
        }
    }
    
    bool compare(const vector<int> &v1, const vector<int> &v2, int i, int j) {
        int lv1 = v1.size(), lv2 = v2.size();
        while (i < lv1 && j < lv2 && v1[i] == v2[j]) {
            i++;
            j++;
        }
        return j == lv2 || (i < lv1 && v1[i] > v2[j]);
    }

    vector<int> maxNumber(vector<int>& nums1, vector<int>& nums2, int k) {
        vector<int> res;
        if (k == 0) return res;
        int ls1 = nums1.size(), ls2 = nums2.size();
        vector<int> v1, v2, v3;
        for (int i = max(0, k - ls2); i <= min(k, ls1); i++) {
            v1.clear(); v2.clear();
            maxkseq(nums1, v1, i);
            maxkseq(nums2, v2, k-i);
            merge(v1, v2, v3);
            if (compare(v3, res, 0, 0))
                res = v3;
        }
        return res;
    }
};

File: python/321. Create Maximum Number.py
#!usr/bin/env python  
#-*- coding:utf-8 _*- 

class Solution(object):
    def maxNumber(self, nums1, nums2, k):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :type k: int
        :rtype: List[int]
        """
        def pick_max(nums, k):
            '''
                保持相对顺序，取nums中k个元素组成最大数字
            '''
            st = []
            drop = len(nums) - k
            for num in nums:
                while drop and st and st[-1] < num:
                    st.pop()
                    drop -= 1
                st.append(num)
            return st[:k]
        
        def merge(A, B):
            '''
            和并A B数组
            '''
            ans = []
            while A or B:
                bigger = A if A > B else B
                ans.append(bigger.pop(0))
            return ans
        
        return max(merge(pick_max(nums1, i), pick_max(nums2, k - i)) for i in range(k+1) if i <= len(nums1) and k - i <= len(nums2))

def pick_max(nums, k):
    '''
        保持相对顺序，取nums中k个元素组成最大数字
    '''
    st = []
    drop = len(nums) - k
    for num in nums:
        while drop and st and st[-1] < num:
            st.pop()
            drop -= 1
        st.append(num)
    return st[:k]


if __name__ == '__main__':
    print pick_max([3, 4, 6, 5], 3)